src/SessionSetup/ServiceDeployManager.cpp

```diff
@@ -449,78 +452,106 @@ ErrorMessageOr<void> ServiceDeployManager::CopyOrbitUserSpaceInstrumentationLibr
+    orbit_service_task_->Write(absl::StrFormat("%s\n", config.root_password));
+    // TODO(antonrohr) Check whether the password was incorrect.
+    // There are multiple ways of doing this. the best way is probably to have a
+    // second task running before OrbitService that sets the SUID bit. It might be
+    // necessary to close stdin by sending EOF, since sudo would ask for trying to
+    // enter the password again. Another option is to use std err as soon as its
+    // implemented in OrbitSshQt::Task.
+  }
 
   orbit_qt_utils::EventLoop loop{};
   auto error_handler =
       ConnectErrorHandler(&loop, &orbit_service_task_.value(), &orbit_ssh_qt::Task::errorOccurred);
-  auto quit_handler =
-      ConnectQuitHandler(&loop, &orbit_service_task_.value(), &orbit_ssh_qt::Task::started);
   auto cancel_handler = ConnectCancelHandler(&loop, this);
+  QObject::connect(
+      &orbit_service_task_.value(), &orbit_ssh_qt::Task::finished, &loop, [&loop](int exit_code) {
+        // TODO(http://b/221369463): Also report a potential error message that has been logged to
+        // stdout.
+        loop.error(ErrorMessage{
+            absl::StrFormat("The service exited prematurely with exit code %d.", exit_code)});
+      });
+
+  std::string stdout_buffer;
+
+  QObject::connect(&orbit_service_task_.value(), &orbit_ssh_qt::Task::readyReadStdOut, &loop,
+                   [&]() {
+                     // We are looking for the kReadyKeyword. Since it might be split up into
+                     // consecutive chunks in the stdout stream we reassemble to whole string into a
+                     // buffer and check that for the keyword.
+                     stdout_buffer.append(orbit_service_task_->ReadStdOut());
+
+                     // That's what we expect the service to send through stdout when it's ready to
+                     // accept a connection from the client.
+                     constexpr std::string_view kReadyKeyword = "READY";
+
+                     if (absl::StrContains(stdout_buffer, kReadyKeyword)) {
+                       ORBIT_LOG("The service reported to be ready to accept connections.");
+                       loop.quit();
+                       return;
+                     }
+
+                     // This is protecting us against consuming unreasonable amount of memory when
+                     // for whatever reason there is a lot of data coming through the stdout
+                     // channel.
+                     constexpr size_t kMaxBufferSize = 100ul * 1024;  // 100 KiB
+
+                     if (stdout_buffer.size() > kMaxBufferSize) {
+                       const auto number_of_bytes_to_remove =
+                           static_cast<ptrdiff_t>(stdout_buffer.size() - kMaxBufferSize);
+                       stdout_buffer.erase(stdout_buffer.begin(),
+                                           stdout_buffer.begin() + number_of_bytes_to_remove);
+                     }
```

Comment:
Since we are currently not using stdout on OrbitService, a lot of data coming in here, probably means something wrong is going on, correct? Your idea is to deliberately not throw an error here, but instead have an error when the timeout occurs, correct?

Reply:
There are multiple aspects here:

1. In the next PR we will make use of this buffer by extracting an error message in case OrbitService fails to start. That's why we need to buffer all the text that's coming in. 
2. But for now, there is only the detection of the READY keyword and the timeout as a fallback.
3. We don't have full control on what data is coming through stdout. The SSH server will spawn a shell which will start OrbitService. It's possible that the shell produces some data for whatever reason. I could imagine a lot of error messages when the hard disk got full or something like that. That's why I wanted to limit the buffer size.

Reply:
If an upcoming PR sends error message, we can talk about it there. stdout vs stderr is also interesting then. I would say, leave it as is in this PR